components/formats-bsd/src/loci/formats/in/OBFReader.java

```diff
@@ -561,131 +651,353 @@ private String readString() throws IOException {
+      if (flushBlockIndex > 0) {
+        seekDestination = stack.flushPoints.get(flushBlockIndex - 1);
+        extraSkipBytes -= flushBlockIndex * stack.flushBlockSize;
+      }
+    }
 
-                in.read(input, 0, length);
-                inflater.setInput(input, 0, length);
-              }
-              else if (isFLIM && remainder == 0 && currentInflatedFrame.number >= 0) {
-                offset = bytes.length;
-                continue;
-              }
-              else {
-                throw new FormatException("Corrupted zlib compression");
-              }
-            }
-            else if (inflater.needsDictionary()) {
-              throw new FormatException("Unsupported zlib compression");
-            }
-            try {
-              offset += inflater.inflate(bytes, offset, bytes.length - offset);
-            }
-            catch (DataFormatException exception) {
-              throw new FormatException(exception.getMessage());
-            }
-          }
-          ++ currentInflatedFrame.number;
-        }
+    if (stack.compression) {
+      state.inflater = new Inflater(seekDestination != 0);
+
+      if (state.inflateInputBuffer == null) {
+        state.inflateInputBuffer = new byte[8192];
       }
-      if (isFLIM) {
-        for (int yy=y; yy<y+h; yy++) {
-          for (int xx=x; xx<x+w; xx++) {
-            int src = getSizeZ() * bytesPerPixel * ((yy * getSizeX()) + xx) + no * bytesPerPixel;
-            int dest = ((yy - y) * w * bytesPerPixel) + ((xx - x) * bytesPerPixel);
-            System.arraycopy(bytes, src, buffer, dest, bytesPerPixel);
-          }
-        }
+    }
+
+    if (!hasChunks) {
+      in.seek(stack.position + seekDestination);
+      skipBytes(stack, extraSkipBytes);
+      return;
+    }
+
+    switchChunk(stack, Collections.binarySearch(
+      stack.chunkLogicalPositions, seekDestination));
+
+    in.seek(state.chunkFileStart +
+      seekDestination - state.chunkLogicalStart);
+
+    skipBytes(stack, extraSkipBytes);
+  }
+
+  private void skipBytes(Stack stack, long byteCount)
+    throws IOException, FormatException
+  {
+    boolean hasChunks = stack.chunkLogicalPositions != null;
+
+    if (!stack.compression && !hasChunks) {
+      in.seek(in.getFilePointer() + byteCount);
+      state.nextReadPosition += byteCount;
+    } else if (stack.compression) {
+      if (skipBuffer == null) {
+        skipBuffer = new byte[8192];
       }
-      else {
-        for (int row = 0; row != h; ++ row) {
-          System.arraycopy(bytes, ((row + y) * columns + x) * bytesPerPixel, buffer, row * w * bytesPerPixel, w * bytesPerPixel);
+
+      while (byteCount > 0) {
+        int readSize = (int)Math.min(skipBuffer.length, byteCount);
+        readFromStack(stack, skipBuffer, 0, readSize);
+        byteCount -= readSize;
+      }
+    } else {
+      long remainingBytesInChunk = remainingBytesInChunk(stack);
```

Comment:
From here through line 845 is nearly identical to the logic in ```readFromStackRaw```.  Can this be refactored to reduce duplicate code?

Reply:
Thanks, I changed `readFromStackRaw` to handle both cases.